src/lib.rs

```diff
@@ -1,3 +1,162 @@
-#[test]
-fn it_works() {
+#[cfg(test)]
+mod test {
+    use tokens::{ Lexer, LexerOptions };
+
+    #[test]
+    fn lexer_usage() {
+        // build the lexer once for project environment with extensions.
+        let lexer = Lexer::new(LexerOptions::default(), vec![]);
+
+        // use many times.
+        {
+            let source: String = "{{ var }}".into();
+            for token in lexer.tokens(&source) {
+                println!("{:?}", token);
+            }
+        }
+    }
+}
+
+// I don't know where to put this, keeping it in root for now.
+#[derive(Debug, Default, Clone)]
+pub struct Position {
+    pub line: usize,
+    pub column: usize,
+}
+
+// Named this "tokens", in plural, to mean a place where you should expect
+// to find your tokens.
+// Similar convention would work for "nodes", that's where AST lives, and possibly
+// "instructions", if we decide to go that route.
+pub mod tokens {
```

Comment:
I think TwigPHP has chosen a very weird name "environment" ... what's that supposed to mean? Does it simulate global warming? No. ;-)

What TwigPHP calls "environment", I prefer to call compiler in a broad sense. With my Wikipedia-knowledge about compilers I would choose the naming like this: The compiler consists of a compiler-frontend (=lexer/parser, analyzing the source code - building an internal representation). What TwigPHP calls "compiler", I would suggest to call a compiler-backend (=optimizer/code generator) - again based on some Wikipedia-knowledge. ;-)

So, if I write compiler, I probably mean Twig-environment. What do you think about this renaming environment -> compiler? Does it make things clearer from a "consumers" point of view?

My motivation behind my crate-layout was this: The "consumer" of this library should be able to either
a) choose a ready-to-use compiler with default-configuration
b) or create a customized compiler by composition of its building blocks (custom lexer, custom parser, ..)

That's why I found it natural to make the "building blocks" of the compiler the "consumer-facing modules".

I think a consumer of this library, will probably not be very much interested about node representations etc., but more about how to configure and tweak the whole thing and its parts, injecting extensions here and there.

You are right, that the parser reaches into the guts of the lexer, in my current layout. I was considering to reverse this - it might make more sense if every part defines its input, something like "_you_ tell me what _you_ need _I_ tell you, what _I_ need"-principle instead of current "I tell you what you need and you tell me what I need"-principle (=lexer, parser, etc. talking with each other on their first date). But that would just revert the direction - then lexer would need to look into "details" of parser.

From the building-blocks perspective of a compiler - which I assume to be the perspective of a library consumer - parser / lexer / loader / runtime / ... are probably more important details than nodes / tokens / ... I am not 100% sure - just trying to take this perspective.